File: backend/app/model/company.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional, TYPE_CHECKING
from sqlmodel import SQLModel, Field, Relationship

if TYPE_CHECKING:
    from .stock import StockData, IntradayTick, DailyOHLC
    from .portfolio import PortfolioPosition, WatchlistItem
    from .alert import Alert, StockNews
    from .trade import Trade
    from .order import Order
    from .fundamental import (
        DividendInformation,
        FinancialPerformance,
        QuarterlyPerformance,
        ShareholdingPattern,
        LoanStatus
    )


class Company(SQLModel, table=True):
    """Company master table - consolidated from old company and stockcompany tables"""
    # Primary identifier (UUID for consistency with rest of system)
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    
    # Basic company information
    name: str = Field(max_length=255)
    trading_code: str = Field(max_length=50, unique=True, index=True)  # Also known as symbol
    company_name: Optional[str] = Field(default=None, max_length=255)  # Full company name
    scrip_code: Optional[int] = None
    
    # Classification
    type_of_instrument: Optional[str] = Field(default=None, max_length=50)
    market_category: Optional[str] = Field(default=None, max_length=1)
    electronic_share: Optional[str] = Field(default=None, max_length=1)
    sector: Optional[str] = Field(default=None, max_length=255, index=True)
    industry: Optional[str] = Field(default=None, max_length=100)
    
    # Capital structure
    authorized_capital: Optional[float] = None
    paid_up_capital: Optional[float] = None
    face_value: Optional[float] = None
    market_lot: Optional[int] = None
    total_outstanding_securities: Optional[int] = None
    total_shares: Optional[int] = None
    free_float: Optional[Decimal] = None
    
    # Market metrics
    market_cap: Optional[Decimal] = None
    pe_ratio: Optional[Decimal] = None
    pb_ratio: Optional[Decimal] = None
    eps: Optional[Decimal] = None
    nav: Optional[Decimal] = None
    dividend_yield: Optional[Decimal] = None
    reserve_and_surplus: Optional[float] = None
    
    # Contact and location
    address: Optional[str] = Field(default=None, max_length=2500)
    factory_address: Optional[str] = Field(default=None, max_length=255)
    phone: Optional[str] = Field(default=None, max_length=255)
    fax: Optional[str] = Field(default=None, max_length=255)
    email: Optional[str] = Field(default=None, max_length=255)
    website: Optional[str] = Field(default=None, max_length=255)
    
    # Company secretary
    company_secretary_name: Optional[str] = Field(default=None, max_length=255)
    company_secretary_email: Optional[str] = Field(default=None, max_length=255)
    company_secretary_cell_no: Optional[str] = Field(default=None, max_length=255)
    
    # Dates and status
    debut_trading_date: Optional[str] = None  # Use str for date, parse as needed
    listing_year: Optional[int] = None
    year_end: Optional[str] = Field(default=None, max_length=25)
    last_agm_date: Optional[str] = None  # Use str for date, parse as needed
    fifty_two_weeks_moving_range: Optional[str] = Field(default=None, max_length=100)
    
    # Status and timestamps
    is_active: bool = Field(default=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Relationships
    stock_data: list["StockData"] = Relationship(back_populates="company")
    intraday_ticks: list["IntradayTick"] = Relationship(back_populates="company")
    daily_data: list["DailyOHLC"] = Relationship(back_populates="company")
    positions: list["PortfolioPosition"] = Relationship(back_populates="stock")
    watchlist_items: list["WatchlistItem"] = Relationship(back_populates="stock")
    trades: list["Trade"] = Relationship(back_populates="stock")
    alerts: list["Alert"] = Relationship(back_populates="stock")
    news_relations: list["StockNews"] = Relationship(back_populates="stock")
    orders: list["Order"] = Relationship(back_populates="stock")


# Pydantic models for API
class CompanyBase(SQLModel):
    """Base model for Company"""
    name: str
    trading_code: str
    company_name: Optional[str] = None
    scrip_code: Optional[int] = None
    type_of_instrument: Optional[str] = None
    market_category: Optional[str] = None
    electronic_share: Optional[str] = None
    sector: Optional[str] = None
    industry: Optional[str] = None
    authorized_capital: Optional[float] = None
    paid_up_capital: Optional[float] = None
    face_value: Optional[float] = None
    market_lot: Optional[int] = None
    total_outstanding_securities: Optional[int] = None
    total_shares: Optional[int] = None
    free_float: Optional[Decimal] = None
    market_cap: Optional[Decimal] = None
    pe_ratio: Optional[Decimal] = None
    pb_ratio: Optional[Decimal] = None
    eps: Optional[Decimal] = None
    nav: Optional[Decimal] = None
    dividend_yield: Optional[Decimal] = None
    is_active: bool = True


class CompanyCreate(CompanyBase):
    """Create model for Company"""
    pass


class CompanyUpdate(SQLModel):
    """Update model for Company"""
    name: Optional[str] = None
    company_name: Optional[str] = None
    scrip_code: Optional[int] = None
    type_of_instrument: Optional[str] = None
    market_category: Optional[str] = None
    electronic_share: Optional[str] = None
    sector: Optional[str] = None
    industry: Optional[str] = None
    authorized_capital: Optional[float] = None
    paid_up_capital: Optional[float] = None
    face_value: Optional[float] = None
    market_lot: Optional[int] = None
    total_outstanding_securities: Optional[int] = None
    total_shares: Optional[int] = None
    free_float: Optional[Decimal] = None
    market_cap: Optional[Decimal] = None
    pe_ratio: Optional[Decimal] = None
    pb_ratio: Optional[Decimal] = None
    eps: Optional[Decimal] = None
    nav: Optional[Decimal] = None
    dividend_yield: Optional[Decimal] = None
    is_active: Optional[bool] = None


class CompanyPublic(CompanyBase):
    """Public model for Company"""
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

File: dse-data-scrapper/src/main/java/com/dohatec/oms/datascraper/model/Company.java
package com.dohatec.oms.datascraper.model;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Entity representing a company in the DSE (Dhaka Stock Exchange)
 * Maps to the 'company' table in the database
 * Exact mapping to database schema
 */
@Getter
@Setter
@Entity
@Table(name = "company")
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Company {

    @Id
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "trading_code", nullable = false)
    private String tradingCode;

    @Column(name = "scrip_code")
    private Integer scripCode;

    @Column(name = "type_of_instrument")
    private String typeOfInstrument;

    @Column(name = "market_category")
    private String marketCategory;

    @Column(name = "electronic_share")
    private String electronicShare;

    @Column(name = "authorized_capital")
    private Double authorizedCapital;

    @Column(name = "paid_up_capital")
    private Double paidUpCapital;

    @Column(name = "face_value")
    private Double faceValue;

    @Column(name = "market_lot")
    private Integer marketLot;

    @Column(name = "total_outstanding_securities")
    private Integer totalOutstandingSecurities;

    @Column(name = "sector")
    private String sector;

    @Column(name = "debut_trading_date")
    private String debutTradingDate;

    @Column(name = "listing_year")
    private Integer listingYear;

    @Column(name = "address")
    private String address;

    @Column(name = "factory_address")
    private String factoryAddress;

    @Column(name = "phone")
    private String phone;

    @Column(name = "fax")
    private String fax;

    @Column(name = "email")
    private String email;

    @Column(name = "website")
    private String website;

    @Column(name = "company_secretary_name")
    private String companySecretaryName;

    @Column(name = "company_secretary_email")
    private String companySecretaryEmail;

    @Column(name = "company_secretary_cell_no")
    private String companySecretaryCellNo;

    @Column(name = "reserve_and_surplus")
    private Double reserveAndSurplus;

    @Column(name = "year_end")
    private String yearEnd;

    @Column(name = "last_agm_date")
    private String lastAgmDate;

    @Column(name = "fifty_two_weeks_moving_range")
    private String fiftyTwoWeeksMovingRange;

    // Additional merged fields from stockcompany
    @Column(name = "company_name")
    private String companyName;

    @Column(name = "industry")
    private String industry;

    @Column(name = "market_cap")
    private BigDecimal marketCap;

    @Column(name = "total_shares")
    private Integer totalShares;

    @Column(name = "free_float")
    private BigDecimal freeFloat;

    @Column(name = "pe_ratio")
    private BigDecimal peRatio;

    @Column(name = "pb_ratio")
    private BigDecimal pbRatio;

    @Column(name = "eps")
    private BigDecimal eps;

    @Column(name = "nav")
    private BigDecimal nav;

    @Column(name = "dividend_yield")
    private BigDecimal dividendYield;

    @Column(name = "is_active")
    private Boolean isActive;

    @Column(name = "is_dsex")
    private Boolean isDsex;

    @Column(name = "is_ds30")
    private Boolean isDs30;

    @Column(name = "is_dses")
    private Boolean isDses;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (updatedAt == null) {
            updatedAt = LocalDateTime.now();
        }
        if (isActive == null) {
            isActive = true;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    @Override
    public String toString() {
        return "Company{" +
                "id=" + id +
                ", tradingCode='" + tradingCode + '\'' +
                ", name='" + name + '\'' +
                ", sector='" + sector + '\'' +
                '}';
    }
}
